### ISSUE TITLE
multinic master does not work properly

### ISSUE THREAD

**Issue by jorgemoralespou** [openshift/origin#2248]
I have a libvirt setup on a linux where each master/node is using 2 NICs/IPs
Adapter one is a Bridged Adapter (eth0) and gets a DHCP address from my broadband router with access to the internet (192.168.121.x) .
Adapter two is a host only adapter (eth1) that I use for the beta3 addressing 192.168.133.x

I have tried many configurations, but get critical problems when it comes to the router deployment. It never works.
I've tried hacking the DNSIP in the master and node-config.yaml files (changing it from the adapter one to adapter two addressing) but it doesn't seem to work. Deleting all pods/services/routes etc doesn't help either.

I've not tried changing the adapters around so the first device picked up is for the 192.168.133.x range. Will that work or does OSE beta3 simply not like multiple NICs? 

When creating a router, I get  following output:

```
[root@ose3-master ~]# osc get pods | awk '{print $1"\t"$3"\t"$5"\t"$7"\n"}' | column -t
POD CONTAINER(S) HOST STATUS
router-2-9bbvr router ose3-master.example.com/192.168.121.121 Pending
```

Instead of:

```
[root@ose3-master ~]# osc get pods | awk '{print $1"\t"$3"\t"$5"\t"$7"\n"}' | column -t
POD CONTAINER(S) HOST STATUS
router-2-9bbvr router ose3-master.example.com/192.168.133.2 Pending
```

My network config is:

```
1: lo    inet 127.0.0.1/8 scope host lo\       valid_lft forever preferred_lft forever
2: eth0    inet 192.168.121.121/24 brd 192.168.121.255 scope global dynamic eth0\       valid_lft 3079sec preferred_lft 3079sec
3: eth1    inet 192.168.133.2/24 brd 192.168.133.255 scope global eth1\       valid_lft forever preferred_lft forever
9: lbr0    inet 10.1.0.1/24 scope global lbr0\       valid_lft forever preferred_lft forever
```

**Comment by rajatchopra** [comment]
Whats your /etc/hosts saying? golang's net.LookupIP will fetch from there first.

**Comment by jorgemoralespou** [comment]
It has same content as the hosts in the beta3 folder of the training.

**Comment by rajatchopra** [comment]
Can you run this program on the master, and see if the IP address is correctly resolved.
http://fpaste.org/222467/17261831/
sudo go run checkip.go ose3-master.example.com
The first IP of the list is what node is considered bound to.

**Comment by jorgemoralespou** [comment]
I have run the go program, and this is what I get:

````
 [root@ose3-master beta3]# go run checkIP.go ose3-master.example.com
[192.168.133.2] : <nil>
````
As you see, there is only one IP resolved back to the DNS name. I get the same result on every box.

**Comment by rajatchopra** [comment]
Ok I tried this with a virtualbox (will try with libvirt next) setup with two interfaces, eth0 (for internet) and eth1(meant to be used intra-cluster), with the hostnames resolving to the IP address assigned to eth1.
I get correct addresses all the way through.

To debug what is possibly wrong with your setup, can you pass the command line args that openshift is running with? Also when you created the master certs with openshift  admin create-master-certs command, what were the args that you provided?

**Comment by jorgemoralespou** [comment]
Hi,
It seems as it is only a display problem, as it seems to be working. 
I have followed training/beta3 steps, so all commands are those shown there.

I have to admit, that disabling NetworkManager was crap, and I had to do a lot of fixing to have it working.

openshift-master.service
````
/usr/bin/openshift start master --images=registry.access.redhat.com/openshift3_beta/ose-${component}:${version} --config=/etc/openshift/master.yaml --loglevel=4
````

openshift-node.service
````
/usr/bin/openshift start node --images=registry.access.redhat.com/openshift3_beta/ose-${component}:${version} --kubeconfig=/var/lib/openshift/openshift.local.certificates/admin/.kubeconfig --loglevel=4 --config=/etc/openshift/node/node-config.yaml
````

master.yaml
````
[root@ose3-master ~]# cat /etc/openshift/master.yaml 
apiVersion: v1
assetConfig:
  logoutURL: ""
  masterPublicURL: https://ose3-master.example.com:8443
  publicURL: https://ose3-master.example.com:8443/console/
  servingInfo:
    bindAddress: 0.0.0.0:8443
    certFile: /var/lib/openshift/openshift.local.certificates/master/server.crt
    clientCA: ""
    keyFile: /var/lib/openshift/openshift.local.certificates/master/server.key
corsAllowedOrigins:
- 127.0.0.1
- localhost
- ose3-master.example.com:8443
dnsConfig:
  bindAddress: 0.0.0.0:53
etcdClientInfo:
  ca: /var/lib/openshift/openshift.local.certificates/ca/cert.crt
  certFile: /var/lib/openshift/openshift.local.certificates/master/etcd-client.crt
  keyFile: /var/lib/openshift/openshift.local.certificates/master/etcd-client.key
  urls:
  - https://ose3-master.example.com:4001
etcdConfig:
  address: ose3-master.example.com:4001
  peerAddress: ose3-master.example.com:7001
  peerServingInfo:
    bindAddress: 0.0.0.0:7001
    certFile: /var/lib/openshift/openshift.local.certificates/etcd/server.crt
    clientCA: /var/lib/openshift/openshift.local.certificates/ca/cert.crt
    keyFile: /var/lib/openshift/openshift.local.certificates/etcd/server.key
  servingInfo:
    bindAddress: 0.0.0.0:4001
    certFile: /var/lib/openshift/openshift.local.certificates/etcd/server.crt
    clientCA: /var/lib/openshift/openshift.local.certificates/ca/cert.crt
    keyFile: /var/lib/openshift/openshift.local.certificates/etcd/server.key
  storageDirectory: /var/lib/openshift/openshift.local.etcd
etcdStorageConfig:
  kubernetesStorageVersion: v1beta3
  openShiftStorageVersion: v1beta1
imageConfig:
  format: openshift3_beta/ose-${component}:${version}
  latest: false
kind: MasterConfig
kubeletClientInfo:
  ca: /var/lib/openshift/openshift.local.certificates/ca/cert.crt
  certFile: /var/lib/openshift/openshift.local.certificates/master/kubelet-client.crt
  keyFile: /var/lib/openshift/openshift.local.certificates/master/kubelet-client.key
  port: 10250
kubernetesMasterConfig:
  masterIP: ""
  schedulerConfigFile: "/etc/openshift/scheduler.json"
  servicesSubnet: 172.30.17.0/24
  staticNodeNames: []
masterClients:
  deployerKubeConfig: /var/lib/openshift/openshift.local.certificates/openshift-deployer/.kubeconfig
  kubernetesKubeConfig: /var/lib/openshift/openshift.local.certificates/kube-client/.kubeconfig
  openshiftLoopbackKubeConfig: /var/lib/openshift/openshift.local.certificates/openshift-client/.kubeconfig
masterPublicURL: https://ose3-master.example.com:8443
oauthConfig:
  assetPublicURL: https://ose3-master.example.com:8443/console/
  grantConfig:
    method: auto
  identityProviders:
  - challenge: true
    login: true
    name: apache_auth
    provider:
      apiVersion: v1
      file: /etc/openshift-passwd
      kind: HTPasswdPasswordIdentityProvider
  masterPublicURL: https://ose3-master.example.com:8443
  masterURL: https://ose3-master.example.com:8443
  sessionConfig:
    sessionMaxAgeSeconds: 300
    sessionName: ssn
    sessionSecretsFile: ""
  tokenConfig:
    accessTokenMaxAgeSeconds: 86400
    authorizeTokenMaxAgeSeconds: 300
policyConfig:
  bootstrapPolicyFile: /var/lib/openshift/openshift.local.policy/policy.json
  masterAuthorizationNamespace: master
  openshiftSharedResourcesNamespace: openshift
servingInfo:
  bindAddress: 0.0.0.0:8443
  certFile: /var/lib/openshift/openshift.local.certificates/master/server.crt
  clientCA: /var/lib/openshift/openshift.local.certificates/ca/cert.crt
  keyFile: /var/lib/openshift/openshift.local.certificates/master/server.key
````
I didn't explicitly create any certificate. Maybe ansible did it for me :-(

**Comment by rajatchopra** [comment]
Ok, something wrong with ansible then, the master certs are usually created with a hostnames arg e.g.
/usr/bin/openshift admin create-master-certs \
    --cert-dir=${CERT_DIR} --hostnames=${MASTER_IP},${MASTER_NAME}

I am guessing that MASTER_IP didnt match the right one on MASTER_NAME. Will try with ansible and see if that is what happens.

Meanwhile, do you have improvement ideas/suggestions on what caused problems with NetworkManager? It is asked that we turn it off because the SDN creates new devices/bridges that should not be managed by NM. We can work around by having the ifcfg- files contain NM_CONTROLLED=no, but I am curious on the hassle that you faced.

**Comment by detiber** [comment]
The beta3 branch of openshift-ansible creates the certificates when generating the master config using openshift start master --write-config, it also sets the hostnames and IPs based on certain detected values as follows:
- if a cloud provider is detected (aws, gce, openstack), then openshift.common.hostname, openshift.common.public_hostname, openshift.common.ip, and openshift.common.public_ip are defaulted to the values derived from the metadata.
- if a cloud provider is not detected, then openshift.common.ip and openshift.common.public_ip are defaulted to the value of the ansible_default_ipv4 fact, which is derived from the ip address associated with the default gateway for the host. openshift.common.hostname and openshift.common.public_hostname are derived from hostname -f, ansible_nodename or ansible_fqdn (depending on the first to return a non-local url that looks fqdn'ish).
- openshift_hostname, openshift_public_hostname, openshift_ip, and openshift_public_ip will override the assocated defaults when configured for a host.

**Comment by jorgemoralespou** [comment]
It seems like the problem is in the ansible installer, since the default gateway, in my setup was the ip used for internet access and not meant for Openshift setup.
I guess that setting up /etc/ansible/hosts with the appropriate values can solve the issue. 
Then, what is needed is proper documentation (if not available) I haven't seen it.

Also, how could I resinstall the nodes after changing this values? If I rerun ansible with modified values, will that work, or will keep the original values?

On the NM side, what I experienced is that it was not easy to modify configuration for one of the NICs. I had to install back NetworkManager (not enabled) and use nmcli command to alter configuration. Not sure why ti worked, though.
